packages/portalnetwork/src/wire/utp/Socket/congestionControl.ts

```diff
@@ -0,0 +1,110 @@
+import { DEFAULT_PACKET_SIZE, MAX_CWND_INCREASE_PACKETS_PER_RTT } from '../index.js'
+import { EventEmitter } from 'events'
+import { Debugger } from 'debug'
+import debug from 'debug'
+
+const CCONTROL_TARGET = 100
+
+export class CongestionControl extends EventEmitter {
+  writing: boolean
+  logger: Debugger
+  rtt: number
+  rtt_var: number
+  timeout: number
+  timeoutCounter?: NodeJS.Timeout
+  baseDelay: { delay: number; timestamp: number }
+  ourDelay: number
+  max_window: number
+  cur_window: number
+  reply_micro: number
+  sendRate: number
+  outBuffer: Map<number, number>
+
+  constructor() {
+    super()
+    this.writing = false
+    this.logger = debug('utp:congestionControl')
+    this.max_window = DEFAULT_PACKET_SIZE * 3
+    this.cur_window = 0
+    this.reply_micro = 0
+    this.rtt = 1000
+    this.rtt_var = 0
+    this.timeout = 1000
+    this.baseDelay = { delay: 0, timestamp: 0 }
+    this.ourDelay = 0
+    this.sendRate = 0
+    this.outBuffer = new Map()
+  }
+
+  async canSend(): Promise<boolean> {
```

Comment:
Just thinking this through.  Should we have some sort of timeout on this so if the `canSend` event isn't emitted within say 3 seconds, we abort the promise and close the socket?  Seems like a memory leak waiting to happen if we leave these promises just sitting waiting if the remote peer doesn't ACK some of the in flight packets because it hangs up/goes offline/etc.